mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -979,7 +979,9 @@ struct FoldEmptyTensorWithDimOp : public OpRewritePattern<DimOp> {
     auto emptyTensorOp = dimOp.getSource().getDefiningOp<EmptyOp>();
     if (!emptyTensorOp || !maybeConstantIndex)
       return failure();
-    if (!emptyTensorOp.getType().isDynamicDim(*maybeConstantIndex))
+    auto emptyTensorType = emptyTensorOp.getType();
```

Comment:
This is fine, but can you provide a repro. for this. Also this doesnt seem to be an error here, but rather a caller error AFAICS.

Reply:
Hi @MaheshRavishankar thanks for your review!

Here is the original repro.

`mlir-opt --linalg-fold-unit-extent-dims a.mlir`

a.mlir:

```mlir
func.func @func1() -> index {
  %3 = tensor.empty() : tensor<19x15x1xf16>
  %idx10 = index.constant 10
  %dim_162 = tensor.dim %3, %idx10 : tensor<19x15x1xf16>  // index out-of-bounds
  return %dim_162 : index
}
```

From my understanding, OOBs `tensor.dim` operations result in UB but that doesn't mean that the IR is invalid (or does it?). The crash here is unexpected due to the assertion check OOBs accesses in`isDynamicDim`.

Reply:
we should fix the crash, but I would change the verifier of `tensor.dim` to assert this as a verification error. If the `idx10` is truly dynamic then I dont think the crash happens.

Reply:
Okay, so you're suggesting to add `hasVerifier = 1` in ODS, then verify the condition? This doesn't seem to be inline with the spec of `tensor.dim` since an OOBs access is not a verification error but UB.

This comment from the `tensor.dim` op `fold` method seems to also suggest that as well:
```
Out of bound indices produce undefined behavior but are still valid IR.
Don't choke on them.
```

I do agree that if the index is truly dynamic, then this fold shouldn't be possible and the crash doesn't occur.

I may be totally confused here but handling the OOBs check in any of the code related to `tensor.dim` op seems to counter the spec of the op itself.

I found a similar check in the `Bufferization` dialect `FoldDimOfAllocTensorOp` pattern (`mlir/lib/Dialect/Bufferization/IR/BufferizationOps.cpp:329`):

```cpp
if (*maybeConstantIndex < 0 ||
    *maybeConstantIndex >= allocTensorOp.getType().getRank())
  return failure();
```

which also suggests I missed a check for when the index is `< 0`.

Reply:
I dont see any benefit in not flagging in the verifier when a tensor.dim is accessed with index greater than the tensor dim (or negative dim). In dynamic index cases you cant do much, its undefined behavior. If you know this statically though its better to flag in the verifier closer to when it is created, rather than letting it plumb through compilation. In all probability this will lead to a runtime error which would be hard to track down. Flagging it in the verifier on creation is better to diagnose errors since I cant think of a valid reason one would create a `tensor.dim` operation with negative index or index out of bounds.

Reply:
That makes sense. I appreciate your reviews, I'll go ahead and make this change!

Reply:
@MaheshRavishankar after further inspection, I found this little excerpt in the docs about the IR verifier: https://mlir.llvm.org/getting_started/DeveloperGuide/#ir-verifier that states statically verifying an expression such as this one in the verifier should be avoided.


> While it is encouraged to verify as much invariants as possible in order to catch bugs during development as soon as possible, there is some important aspect to keep in mind. In particular a common point of confusion is about how to handle “undefined behavior” cases. For example the tensor.dim operation:

```mlir
// Returns the dimension of %A indexed by %dim.
%y = tensor.dim %A, %dim : memref<4x?xf32>
```

> The %dim indicates what dimension to return. If the dimension index is out of bounds, the behavior is undefined. What about:

```mlir
%ten = arith.constant 10 : index
%y = tensor.dim %A, %ten : memref<4x?xf32>
```

> We have unambiguously a violation of the spec here, and we can statically verify it. However this is not the kind of invariants to enforce in a verifier because it relies on non-local properties, which makes the design of the compiler much less flexible.

Implementing the change in the verifier seems to have other implications in the lit tests here in these files:
`Dialect/MemRef/resolve-dim-ops.mlir` and  `Dialect/Tensor/canonicalize.mlir`. The test cases seem to suggest that even though the indices are statically known they should still pass. 

When implementing the change in the verifier the test cases fail. For example in the function `dim_out_of_bounds_2` in `resolve-dim-ops.mlir`:

```
# .---command stderr------------
# | within split at /llvm-project/mlir/test/Dialect/MemRef/resolve-dim-ops.mlir:13 offset :13:8: error: 'tensor.dim' op out-of-range access, attempted to access index 7 but valid range is [0, 1].
# |   %0 = tensor.dim %alloc, %idx : tensor<?x?xf32>
# |        ^
# | within split at /llvm-project/mlir/test/Dialect/MemRef/resolve-dim-ops.mlir:13 offset :13:8: note: see current operation: %3 = "tensor.dim"(%2, %0) : (tensor<?x?xf32>, index) -> index
# `-----------------------------
# error: command failed with exit status: 1
```

These tests previously passed without the verifier change. Not sure what to make of this so your thoughts are much appreciated!

Reply:
Wow, thats funny that this is actually called out. I dont agree with that too much (maybe I agree with the "non-localness" of things). We have a lot of such examples. Maybe at a time this was useful thumb-rule and there might be instances where this principle would be a good thing to follow, but IMO that doesnt hold.
If that test is failing, I am fine removing it as invalid test.

Reply:
The non-localness is strong principle of verifiers checks. Please don't do this in MLIR upstream ever, and fix any verifier that would violate his rule: it is still as accurate as when it has been written down.

Reply:
Sure, I am not commenting on the larger principle that makes sense. In this particular case I think the `tensor.dim` verifier needs to be extended to catch undefined behavior cause this is essentially the same check being duplicated everywhere to leave undefined behavior through to run time. But I dont want to make this a philosophical discussion. Lets fix the issue at hand.

Reply:
So seems like the resolution is to not put this in the verifier and keep it in the `FoldEmptyTensorWithDimOp` pattern. Let me revert the verifier change.

Reply:
Yeah, that end state makes me sad. Ill keep this in mind to the point where this becomes a runtime error and point to this resolution as the root cause.

Reply:
What "runtime" error are you thinking of?

Reply:
Yes that is true, but when it is constant and you can flag it as UB in the compiler we are essentially chosing not to. So something is accessing dimensions of tensor that is out of bound. This unless caught in a compiler most definitely will lead to run time error on the generated code (either correctness cause you indexing might be off, or out of bounds error).

Reply:
>  when it is constant and you can flag it as UB in the compiler we are essentially chosing not to. 

You can absolutely write a pass that would abort compilation in this case if you want. This just does not belong to the IR verifier because UB isn't invalid IR. 

Your local observation of UB in the program isn't even a sign of an incorrect program, because UB is a runtime construct: that is it is very possible that what you see is dead-code.

For example (pseudo IR):
```
%a = constant .... tensor<10xf32>
if (cond) {
 ... = tensor.dim %a, 2
 ....
}
```

Is only an invalid program if at runtime the condition is true. However this isn't an incorrect program if the code-path is never executed. 
Hence it can't be "invalid IR".

Reply:
(you could though fold all of these cases to a `ub.unreachable()` intrinsic for example):

```
%a = constant .... tensor<10xf32>
if (cond) {
  ...
  ub.unreachable()
  ...
}
```

Reply:
Thank you both for the fruitful and insightful discussion, I've learned a ton from your conversation.

Judging from @MaheshRavishankar comments, there might be other places where non-local properties are being used to verify valid IR?

If that is the case, is this something that should be changed to uphold the non-localness principle? Do you have any examples of when this happens that should be changed? I'm happy to take on the task.

Reply:
If you are going to go through larger changes I would first ask on Discourse. As a user of MLIR such changes and "loosening" of verification can have very disruptive effects. I would strongly recommend that we discuss this on discourse first cause such blanket principles without some more nuance on cases-by-case basis is BAD.

Reply:
> (you could though fold all of these cases to a `ub.unreachable()` intrinsic for example):
> 
> ```
> %a = constant .... tensor<10xf32>
> if (cond) {
>   ...
>   ub.unreachable()
>   ...
> }
> ```

You are missing the point and going on a tangent. The same check is being done in many places.. so your suggestions seems to be
1) Keep doing that. Check at every place if the tensor.dim is out-of-bounds access.
2) Add a different pass that verifies this, but that would have to happen everytime we run the verifier. I dont understand how this is expected to work.

Reply:
I don't understand your point actually: this check only applies to places where you would try to **fold** the constant. This shouldn't happen in so many places.
Regardless, this is something common to every op that has UB on specific input values, that happens all the time!

For example you seems like you'd want to make the LLVM verifier reject this IR (which is runtime UB):

```
%cst1 = llvm.constant -1 : i32
%overflow = llvm.add %0, %0 nsw : i32
```



> If you are going to go through larger changes I would first ask on Discourse. As a user of MLIR such changes and "loosening" of verification can have very disruptive effects. I would strongly recommend that we discuss this on discourse first cause such blanket principles without some more nuance on cases-by-case basis is BAD.

Sorry: I won't. If I find verifiers in MLIR that are violating this principle without a very strong exceptional argument (I know of a couple of these out-of-tree that I don't want to get into here), I **will** remove these without any Discourse thread. We have a policy and guideline on this, I am sorry you don't seem to understand the basic principle of the design behind the IR consistency and validity here, I'm happy to explain it to you further, but I won't jump through these hoops you seem to intend to put here in the way of keeping the code consistent from a system design point of view.

Reply:
I am saying I challenge the principle and we need to look at it. This is a community project. No one person gets to ram through their changes without consensus. Consensus can change with time too

Reply:
Of course, but the current consensus and principle exists for now and so this is what we live with at the moment.. Feel free to challenge it but you have to run through things through first principles, and present strong argument and examples to reverse such policy.